Adaptive queue for staging dials

raulk opened the pull request:
Currently the DHT is performing dials outside of the Alpha concurrency limit. We are dialling all nodes that peers return in `CloserPeers` without limit. As a result, we end up flooding the swarm with dial jobs, which trips over the file descriptor limits, and brings dialling to a halt under some circumstances. Our current approach is also algorithmically incorrect, and leads to suboptimal query patterns.

This patch introduces an adaptive dial queue that spawns a dynamically sized set of goroutines to preemptively stage dials for later handoff to the DHT protocol for RPC. It identifies backpressure on both ends (dial consumers and dial producers), and takes compensating action by adjusting the worker pool.

We start with `DialQueueMinParallelism` number of workers (6), and scale up and down based on demand and supply of dialled peers.

The following events trigger scaling:
- we scale up when we can't immediately return a successful dial to a new consumer.
- we scale down when we've been idle for a while waiting for new dial attempts.
- we scale down when we complete a dial and realise nobody was waiting for it.

Dialler throttling (e.g. FD limit exceeded) is a concern, as we can easily spin up more workers to compensate, and end up adding fuel to the fire. Since we have no deterministic way to detect this for now, we hard-limit concurrency to `DialQueueMaxParallelism` (20).

Testing this patch in a production mirror reduced dial backlog considerably, and showed the adaptiveness in action:

```
Jan 29 00:42:02 localhost ipfs[6248]: 2019-01-29 00:42:02.873850 DEBUG dht dial_queue.go:177: grew dial worker pool: 6 => 9
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.003489 DEBUG dht dial_queue.go:177: grew dial worker pool: 6 => 9
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.083072 DEBUG dht dial_queue.go:177: grew dial worker pool: 9 => 13
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.186067 DEBUG dht dial_queue.go:177: grew dial worker pool: 6 => 9
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.246148 DEBUG dht dial_queue.go:199: shrunk dial worker pool: 9 => 6
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.246174 DEBUG dht dial_queue.go:177: grew dial worker pool: 13 => 19
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.328308 DEBUG dht dial_queue.go:177: grew dial worker pool: 6 => 9
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.451401 DEBUG dht dial_queue.go:177: grew dial worker pool: 6 => 9
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.462604 DEBUG dht dial_queue.go:177: grew dial worker pool: 9 => 13
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.463808 DEBUG dht dial_queue.go:177: grew dial worker pool: 9 => 13
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.464417 DEBUG dht dial_queue.go:177: grew dial worker pool: 9 => 13
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.465973 DEBUG dht dial_queue.go:177: grew dial worker pool: 9 => 13
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.470865 DEBUG dht dial_queue.go:199: shrunk dial worker pool: 18 => 12
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.697266 DEBUG dht dial_queue.go:177: grew dial worker pool: 6 => 9
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.697847 DEBUG dht dial_queue.go:177: grew dial worker pool: 6 => 9
Jan 29 00:42:03 localhost ipfs[6248]: 2019-01-29 00:42:03.838040 DEBUG dht dial_queue.go:177: grew dial worker pool: 9 => 13
```

anacrolix commented on dial_queue.go:
```diff
@@ -0,0 +1,269 @@
+var DialQueueMinParallelism = 6
+var DialQueueMaxParallelism = 20
+var DialQueueMaxIdle = 5 * time.Second
+var DialQueueScalingMutePeriod = 1 * time.Second
```
This pattern in the PL codebases is irksome. I think generally global constants should only be vars if there's no alternative. Also consider `var ( ... )` or `const ( ... )`.

vyzo replied:
we usually do put it in a enclosing var, as it should be done here too.

anacrolix commented on dial_queue.go:
```diff
@@ -0,0 +1,269 @@
+		in:  in,
+		out: queue.NewChanQueue(ctx, queue.NewXORDistancePQ(target)),
+
+		growCh:    make(chan struct{}, nConsumers),
```
Channels seem great at first, but they can be very difficult to use, and aren't actually faster than the alternatives. Dunno if you're familiar with this. A good sign that they're not right is if you still need locks somewhere.

anacrolix commented on dial_queue.go:
```diff
@@ -0,0 +1,269 @@
+	"sync"
+	"time"
+
+	peer "github.com/libp2p/go-libp2p-peer"
```
🤢

raulk commented:
Future optimisation: cancelling pending dials to worse nodes as we find closer nodes to the target.

Stebalien commented on dial_queue.go:
```diff
@@ -0,0 +1,269 @@
+
+	// we have no finished dials to return, trigger a scale up.
+	select {
+	case dq.growCh <- struct{}{}:
```
May not be the bet API but it should be considered: Instead of explicitly growing like this, we could:

1. Expose a channel to the user.
2. When we have no new connections, try writing `nil` to the channel.
3. If a consumer reads `nil`, (a) it should throw it away and try again and (b) we should treat that as a signal to grow.

We could also just have the consumer read from *two* channels.

---

Just something to consider. The current API is definitely *cleaner*.

raulk replied:
I did consider consumer-side signalling, but I didn't want to introduce that level of coupling. This fix is really the first iteration of this queue (which could be exportable to other areas of libp2p); we might need to get more sophisticated over time, and making the consumer conform to a specific choreography can be very limiting and end up becoming technical debt in the future.

Stebalien replied:
SGTM.

Stebalien commented on dial_queue.go:
```diff
@@ -0,0 +1,269 @@
+
+		growCh:    make(chan struct{}, nConsumers),
+		shrinkCh:  make(chan struct{}, 1),
+		waitingCh: make(chan waitingCh, nConsumers),
```
Nit: What if we buffered these in a slice inside `control()`? That would get rid of the only real error case in `Consume`.

raulk replied:
I think it boils down to the same, no? We'd allow the queue to grow unboundedly. I prefer us to detect the programming mistake early.

Stebalien replied:
If we have a fixed number of consumers, it's the same. I'm just suggesting we let the consumer dynamically decide (so we don't end up in a weird case where someone ups the number of consumers but forgets to update `nConsumers`) but it doesn't really matter either way.

Stebalien commented on dial_queue.go:
```diff
@@ -0,0 +1,269 @@
+	select {
+	case dq.waitingCh <- waitingCh{ch, time.Now()}:
+	default:
+		return nil, errors.New("detected more consuming goroutines than declared upfront")
```
This should probably just panic.

raulk replied:
If you're fine with that, I am too :)

Stebalien commented on dial_queue.go:
```diff
@@ -0,0 +1,269 @@
+	}
+}
+
+func (dq *dialQueue) Consume() (<-chan peer.ID, error) {
```
Instead of returning an error, maybe return a closed channel to indicate that we're done? It's unclear what do do with the error because it *could* mean "we're done" or it could mean "you did something wrong". Even if we get rid of the second case (panic), this still doesn't really feel like an "error" case.

raulk replied:
I think that's fair.

raulk commented:
Addressed the review comments, but I noticed a flaky test on CI along the way. I do deplore depending on time, but I cannot think of another way to test this.

raulk commented on dial_queue.go:
```diff
@@ -0,0 +1,269 @@
+
+	// park the channel until a dialled peer becomes available.
+	select {
+	case dq.waitingCh <- waitingCh{ch, time.Now()}:
```
The issue with doing that is that we wouldn't be able to detect how many peers are actually waiting to trigger a downscaling: https://github.com/libp2p/go-libp2p-kad-dht/pull/237/files#diff-760f8dba1177a2c3e39fa07a627748e2R254

Stebalien replied:
Ah, I see. Yeah, that makes sense.

raulk commented on query.go:
```diff
@@ -201,14 +204,14 @@ func (r *dhtQueryRunner) spawnWorkers(proc process.Process) {
+			if err != nil {
+				log.Warningf("error while fetching a dialled peer: %v", err)
+				r.rateLimit <- struct{}{}
+				continue
```
What should be the right behaviour be here? An error here could happen if the context is closed.

raulk replied:
Nevermind, I saw the light. Fixed.

raulk commented on query.go:
```diff
@@ -201,14 +204,14 @@ func (r *dhtQueryRunner) spawnWorkers(proc process.Process) {
-					return // channel closed.
-				}
-
+			case p, _ := <-ch:
```
No, and that's a good point. I'm not cleaning up waiting channels when the context is done. Gotta fix that.

raulk commented on query.go:
```diff
@@ -201,14 +204,14 @@ func (r *dhtQueryRunner) spawnWorkers(proc process.Process) {
 		case <-r.rateLimit:
+			ch, err := r.peersDialed.Consume()
+			if err != nil {
+				log.Warningf("error while fetching a dialled peer: %v", err)
```
Removed the second return value (err), and replaced the check with a nil ("", urgh) check on the value. We now return.

Stebalien commented on dial_queue.go:
```diff
@@ -0,0 +1,273 @@
+)
+
+var (
+	DialQueueMinParallelism    = 6
```
Documentation.

Stebalien commented on dial_queue.go:
```diff
@@ -0,0 +1,273 @@
+			if resp.ch != nil {
+				close(resp.ch)
+			}
+			for w := range waiting {
```
Will this ever finish?

raulk replied:
Oh my, I rushed it and that was sloppy of me.

Stebalien commented on query.go:
```diff
@@ -201,14 +204,13 @@ func (r *dhtQueryRunner) spawnWorkers(proc process.Process) {
-					r.rateLimit <- struct{}{}
-					return // channel closed.
+			case p, _ := <-ch:
+				if p == "" {
```
It's slightly cleaner to read `p, ok := <-ch` and then check `ok`.

raulk replied:
👍 For some reason I thought we received the peer and the closure at the same time, but they are different receives.

Stebalien commented on dial_queue.go:
```diff
@@ -91,18 +91,23 @@ func (dq *dialQueue) control() {
+		if resp.ch != nil {
+			close(resp.ch)
+		}
+		close(dq.waitingCh)
```
Is this safe? Aren't we going to get a concurrent `Consume()`?

I think the solution here *may* be to:

1. Make this channel synchronous.
2. Buffer internally (in a slice).

I can't think of a non-racy alternative.

raulk replied:
I thought a select with a default branch, where one of the cases was sending on a closed channel, wouldn't panic. But it does. 

Would setting the `dq.waitingCh = nil` before closing would remove the race?

Regardless, I'm starting to think the buffering in a slice is a good option.

Stebalien replied:
I think the only other way is to lock around writes to `waitingCh` but that gets ugly fast (see the entire go-ipfs commands library).
